### PR TITLE
HPCC-13472 Add new Stored Format(password) option to hide passwords

### PR DESCRIPTION
**Pull request by afishbeck**
Signed-off-by: Anthony Fishbeck anthony.fishbeck@lexisnexis.com

### CODE REVIEW

**Comment by afishbeck** [comment]
@ghalliday please review
@wangkx did I miss anything?
@cloLN care to play hacker, and try and expose a stored password through EclWatch or command line tools?

**Comment by afishbeck** [comment]
see testing/regress/ecl/formatstored.ecl for an example / test case.

**Comment by wangkx** [comment]
in WsWuInfo::getWorkunitQueryShortText()?

**Comment by afishbeck** [comment]
@wangkx I think QueryShortText is ok.  I don't think it's not the default for the password provided in the ECL code that we care about... just if one was provided as an input parameter.

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| formatstored | Pass | Pass | Pass |
 
HPCC Stop: OK

**Comment by wangkx** [comment]
@afishbeck fine for me.

**Comment by cloLN** [comment]
I caused the eclagent to reply with a segfault message (no core file was found).
All I have to do is submit html formatting along with my password from inside the password field and it will be enough to cause the eclagent to segfault.

Steps to recreate:
From wsecl, I used the SOAP form to submit my test,

I submitted the following soap request into the form for the suggested regression test:

```
  <?xml version="1.0" encoding="UTF-8"?>
  <soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:SOAP-ENC="http://schemas.xmlsoap.org/soap/encoding/">
  <soap:Body>
  <formatstoredRequest xmlns="urn:hpccsystems:ecl:formatstored">
   <pw><bold>test</bold></pw>
  </formatstoredRequest>
  </soap:Body>
 </soap:Envelope>

```

**Comment by afishbeck** [comment]
I think you should open a separate JIRA about that.  It seems to pre-exist this change.

**Comment by cloLN** [comment]
NP.  Other than that, this change looks good :)

File: dali/daliadmin/daliadmin.cpp [hunk 1]
```diff
@@ -2417,7 +2417,7 @@ static void dumpWorkunit(const char *wuid, bool includeProgress)
 {
     Owned<IWorkUnitFactory> factory = getWorkUnitFactory();
     Owned<IConstWorkUnit> workunit = factory->openWorkUnit(wuid, false);
-    exportWorkUnitToXMLFile(workunit, "stdout:", 0, true, includeProgress);
+    exportWorkUnitToXMLFile(workunit, "stdout:", 0, true, includeProgress, false);  //should we hide passwords here?
```
Comment (richardkchapman) [inline dali/daliadmin/daliadmin.cpp]: Yes

File: ecl/eclagent/eclagent.cpp [hunk 2]
```diff
@@ -751,7 +751,7 @@ void EclAgent::outputFormattedResult(const char * name, unsigned sequence, bool
     {
     case ofXML:
         {
-            res->getResultXml(buff);
+            res->getResultXml(buff, false);
```
Comment (richardkchapman) [inline ecl/eclagent/eclagent.cpp]: Should be true, I would have thought?

File: ecl/wutest/wutest.cpp [hunk 3]
```diff
@@ -69,7 +69,7 @@ bool dump(IConstWorkUnit &w, IProperties *globals)
     else if (stricmp(action, "dump")==0)
     {
         SCMStringBuffer xml;
-        exportWorkUnitToXML(&w, xml, true, false);
+        exportWorkUnitToXML(&w, xml, true, false, false);
```
Comment (richardkchapman) [inline ecl/wutest/wutest.cpp]: true would be more appropriate I think

File: tools/wuget/wuget.cpp [hunk 4]
```diff
@@ -58,7 +58,7 @@ int main(int argc, char **argv)
                 {
                     Owned<ILocalWorkUnit> wu = createLocalWorkUnit();
                     wu->loadXML(xml);
-                    exportWorkUnitToXML(wu, xml.clear(), true, false);
+                    exportWorkUnitToXML(wu, xml.clear(), true, false, false); //should we hide passwords here?
```
Comment (richardkchapman) [inline tools/wuget/wuget.cpp]: Yes (though it would only reveal the default...)

File: common/workunit/workunit.cpp [hunk 5]
```diff
@@ -3666,7 +3666,7 @@ bool CLocalWorkUnit::archiveWorkUnit(const char *base,bool del,bool ignoredllerr
         return false;
 
     StringBuffer buf;
-    exportWorkUnitToXML(this, buf, false, false);
+    exportWorkUnitToXML(this, buf, false, false, false);
```
Comment (richardkchapman) [inline common/workunit/workunit.cpp]: Should consider hiding pw in archive too I think

Comment (afishbeck) [inline common/workunit/workunit.cpp]: would mean un-archived workunits could no longer run or behave the same?

Comment (richardkchapman) [inline common/workunit/workunit.cpp]: When rerunning a workunit you would need to resupply the password... I'm not sure that is a bad thing.

Comment (afishbeck) [inline common/workunit/workunit.cpp]: Ok sounds right.

**Comment by afishbeck** [comment]
@richardkchapman updated

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| formatstored | Pass | Pass | Pass |
 
HPCC Stop: OK

**Pull request merged**